File: leetcode_journey/Q46_Permutations.py
"""
Given a collection of distinct integers, return all possible permutations.

Example:

Input: [1,2,3]
Output:
[
  [1,2,3],
  [1,3,2],
  [2,1,3],
  [2,3,1],
  [3,1,2],
  [3,2,1]
]
"""

# from itertools import permutations

class Solution(object):
    # #use existing module
    # def permute1(self,nums):
    #     return permutations(nums)


#recursion:
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if len(nums) == 0:
            return []

        if len(nums) ==1 :
            return [nums]
        answer = []
        for i, num in enumerate(nums):
            rest = nums[:i] + nums[i+1:]
            for y in self.permute(rest):
                answer.append([num] + y)

        return answer

#

if __name__ == "__main__":
    a = [1,2,3]
    print (Solution().permute(a))

File: src/Q46_Permutations.java
/*
Given an array nums of distinct integers, return all the possible permutations. You can return the answer in any order.



Example 1:

Input: nums = [1,2,3]
Output: [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
Example 2:

Input: nums = [0,1]
Output: [[0,1],[1,0]]
Example 3:

Input: nums = [1]
Output: [[1]]


Constraints:

1 <= nums.length <= 6
-10 <= nums[i] <= 10
All the integers of nums are unique.
 */

import java.util.List;
import java.util.ArrayList;
public class Q46_Permutations {
    public List<String> permutations(String set){
        List<String> result = new ArrayList<String>();
        //sanity check
        if (set == null){
            return result;
        }
        char[] array = set.toCharArray();
        swap(array, 0, result);
        return result;
    }

    //choose the character to be at the position of "Index"
    //all the already chosen positions are (0, index - 1)
    //all the candidate characters can be at position "index"
    //are in the subarray of (index, array.length-1)
    private void helper(char[] array, int index,  List<String> result){
        //terminate condition:
        //only when we have already chosen the characters for all the positions,
        //we can have a complete permutation
        if (index == array.length){
            result.add(new String(array));
            return;
        }
        //all the possible characters could be placed at index are
        //the characters in the subarray (index, array.length-1)
        for (int i = index; i < array.length; i++){
            swap(array, index, i);
            helper(array, index + 1, result);
            //remember to swap back when back track to previou level
            swap(array, index, 1);
        }
    }

    private void swap(char[] array, int left, int right){
        char tmp = array[left];
        array[left] = array[right];
        array[right] = tmp;

    }

}
